### PR TITLE
[SPARK-17750][SQL] Fix CREATE VIEW with INTERVAL arithmetic.

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Currently, Spark raises `RuntimeException` when creating a view with timestamp with INTERVAL arithmetic like the following. The root cause is the arithmetic expression, `TimeAdd`, was transformed into `timeadd` function as a VIEW definition. This PR fixes the SQL definition of `TimeAdd` and `TimeSub` expressions.

``` scala
scala> sql("CREATE TABLE dates (ts TIMESTAMP)")

scala> sql("CREATE VIEW view1 AS SELECT ts + INTERVAL 1 DAY FROM dates")
java.lang.RuntimeException: Failed to analyze the canonicalized SQL: ...
```
## How was this patch tested?

Pass Jenkins with a new testcase.

### CODE REVIEW

**Comment by rxin** [comment]
Shouldn't the right test be just have a SQL generation test for interval arithmetics?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @rxin !

Do you mean `SQLQueryTestSuite`?
When I tried with it there, it fails because this needs Hive support. 

The other testcase there uses temporary views. But, this bug only occurs on permanent views. When I tried with temporary view, there was no problem. So, I put this here.

**Comment by dongjoon-hyun** [comment]
Oh, Sorry. You mean SQLBuilderSuite, right?
I see. I will do like that.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
I removed the old test and created `test("interval arithmetic")` in `ExpressionSQLBuilderSuite`.

**Comment by rxin** [comment]
this too, but i think we should also add a sql generation test in the sql generation suite.

**Comment by dongjoon-hyun** [comment]
@rxin .
Could you give me the name of the suite what you mean by the following?

> sql generation suite.

I'll add the testcase there, too.

**Comment by gatorsmile** [comment]
`LogicalPlanToSQLSuite`

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile ! I'll add there, too.

**Comment by dongjoon-hyun** [comment]
Oh, I completely forgot about that testsuite. What a shame on me.

**Comment by dongjoon-hyun** [comment]
Thank you, @rxin and @gatorsmile .
Finally, I added that correctly.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala [hunk 1]
```diff
@@ -119,4 +121,18 @@ class ExpressionSQLBuilderSuite extends SQLBuilderTest {
       s"(PARTITION BY `a`, `b` ORDER BY `c` ASC NULLS FIRST, `d` DESC NULLS LAST $frame)"
     )
   }
+
+  test("interval arithmetic") {
+    val interval = Literal(new CalendarInterval(0, 86400000000L))
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala]: Please add a comment about this fly number: `milliseconds per 24 hours`

**Comment by gatorsmile** [comment]
I have a question about the SQL statement: 
```SQL
select ts + interval 1 days, ts - interval 2 days from dates
```

Is it a valid SQL statement? I tried it in Hive and DB2. Both do not accept it.

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala]: Sure!

**Comment by dongjoon-hyun** [comment]
I think you mean 'DAYS' is wrong right? I think the following should work there.
```sql
select ts + interval 1 day, ts - interval 2 day from dates
```

**Comment by gatorsmile** [comment]
```
db2 => select ts + interval 1 days, ts - interval 2 days from dates;
SQL0104N  An unexpected token "1" was found following "select ts + interval".  
Expected tokens may include:  "<space>".  SQLSTATE=42601
db2 => select ts + interval 1 day, ts - interval 2 day from dates
SQL0104N  An unexpected token "1" was found following "select ts + interval".  
Expected tokens may include:  "<space>".  SQLSTATE=42601
```

Also tried it in Hive.

**Comment by dongjoon-hyun** [comment]
There are a few things are mixed.

First of all, when I checked in MySQL. It supports `select ts + interval 1 day, ts - interval 2 day from dates`.
Second, more important, as you see, `days` is the generated string from the current Spark. I think we had better to change that into `day`.

Which version of Hive do you mean? For `INTERVAL`, Spark 1.6.2 and hive 1.2 does not support that.

BTW, I'm willingly to fix anything. :)

**Comment by dongjoon-hyun** [comment]
For Postgres, it works like this.
```
select ts + interval '1' day, ts - interval '2' day from dates
```

**Comment by dongjoon-hyun** [comment]
Maybe, could you try with '1' instead of 1?

**Comment by dongjoon-hyun** [comment]
To sum up, if you agree, I will changed the followings.
- Use `day` instead of `days` (I need to find where it is.)
- Use string type, `1`,  instead of integer type, 1.

**Comment by gatorsmile** [comment]
It sounds this is database vendor-specific. Please try all the variants. If the queries work in Spark SQL, add these queries into your test cases for SQL generation. 

BTW, the PR description can be updated.

**Comment by dongjoon-hyun** [comment]
I see.
> Please try all the variants

Our Apache Spark supports all the following 4 cases.
- interval 1 day
- interval 1 days
- interval '1' day
- interval '1' days

```scala
scala> sql("select current_timestamp + INTERVAL 1 DAY, current_timestamp + INTERVAL 1 DAYS, current_timestamp + INTERVAL '1' DAY, current_timestamp + INTERVAL '1' DAYS").show
+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+
|CAST(current_timestamp() + interval 1 days AS TIMESTAMP)|CAST(current_timestamp() + interval 1 days AS TIMESTAMP)|CAST(current_timestamp() + interval 1 days AS TIMESTAMP)|CAST(current_timestamp() + interval 1 days AS TIMESTAMP)|
+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+
|                                    2016-10-02 14:49:...|                                    2016-10-02 14:49:...|                                    2016-10-02 14:49:...|                                    2016-10-02 14:49:...|
+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+--------------------------------------------------------+
```

**Comment by dongjoon-hyun** [comment]
I'll update the testcases and PR description. Thank you again, @gatorsmile !

**Comment by dongjoon-hyun** [comment]
BTW, Spark accepts the four cases and return one normalized case, 'INTERVAL 1 DAYS'. I think we don't need to change the current behavior here. Right?

**Comment by gatorsmile** [comment]
IMO, you do not need to fix the existing behavior.

Maybe you also can check the original PR that delivered the feature of Temporal Interval and check the design and coverage. I quickly went over the SQL-99. Temporal operations are many. I am not sure whether we cover all the issues.

**Comment by dongjoon-hyun** [comment]
Ya. I agree. We can revisit if the existing behavior has some issue.

**Comment by dongjoon-hyun** [comment]
I addressed the comments. For the PR description, it looks okay because we can not enumerate all cases there with the same reason.

**Comment by gatorsmile** [comment]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you for review!

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Could you review this PR about `INTERVAL` again?

File: sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala [hunk 2]
```diff
@@ -119,4 +121,18 @@ class ExpressionSQLBuilderSuite extends SQLBuilderTest {
       s"(PARTITION BY `a`, `b` ORDER BY `c` ASC NULLS FIRST, `d` DESC NULLS LAST $frame)"
     )
   }
+
+  test("interval arithmetic") {
+    val interval = Literal(new CalendarInterval(0, 86400000000L)) // milliseconds per 24 hours
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala]: Can you use `MICROS_PER_DAY` instead?

Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/catalyst/ExpressionSQLBuilderSuite.scala]: Sure!

**Comment by rxin** [comment]
LGTM, pending Jenkins.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile .
Could you merge this PR? :)

**Comment by rxin** [comment]
Yea @gatorsmile you should try.

**Comment by gatorsmile** [comment]
Will make a try. Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .

**Comment by dongjoon-hyun** [comment]
Thank you, @rxin and @gatorsmile .

**Comment by gatorsmile** [comment]
Merging to master. Thanks!

I think we also need to merge it to 2.0 branch, right?

**Comment by dongjoon-hyun** [comment]
Sure. I'll make abackport for this.

**Comment by gatorsmile** [comment]
Thank you!

**Pull request closed without merging**